new LightningModule hook "configure_callbacks"

awaelchli opened the pull request:
## What does this PR do?

Fixes #5615 

Introduces a `configure_callbacks` hook in LightningModule
Possible implementations:

**Option 1: "appending" (CURRENT IMPLEMENTATION ON THIS PR)**
Simply appens the returned callbacks. Existing callbacks cannot be modified directly.
```python

def configure_callbacks(self):
    my_callback = MyCallback()
    return [my_callback]

# this is what the trainer would do:
trainer.callbacks.extend(model.configure_callbacks())
```

**Option 2: "replacement"**
Allows to define the order and to modify existing callbacks.
```python

def configure_callbacks(self, callbacks):
    my_callback = MyCallback()
    return [callbacks] + callbacks

# this is what the trainer would do:
trainer.callbacks = model.configure_callbacks()
```

**Option 3: "inplace modification"**
Allows to define the order and to modify existing callbacks.
```python

def configure_callbacks(self, callbacks):
    my_callback = MyCallback()
    callbacks.append(my_callback)
    # no return

# this is what the trainer would do:
model.configure_callbacks(trainer.callbacks)
```

Options 2 and 3 give the user more control over how callbacks are used and in which order they are executed. 
However, if the LightningModule modifies the the callbacks passed to the Trainer, it is less transparent what code actually gets executed. On the other hand, a user can mutate the callbacks list in the trainer anyway through `Trainer.callbacks`, if they want to.


TODOs:
- [x] decide on the interface, discussion needed (see proposed options below)
- [x] update docs
- [x] find the best place to invoke the hook
- [x] add tests
- [x] add changelog
- [x] Merge #5731, ensures that model checkpoints will run last

**Poll options**
1. - ❤️ 
2. - 🎉 
3. - 🚀

akihironitta commented:
+1 for **`Option 1: "appending"`** since it is simple and similar to `configure_optimizers()` so that users don't get confused. Also, I'm curious about the motivation for adding this feature...

awaelchli commented:
> I'm curious about the motivation for adding this feature...

This was requested internally, see grid chat. The motivation is to have model-specific callbacks defined in the model, making it more portable.

ananthsub commented:
@awaelchli what are the implications for checkpoint loading/saving? how are callback states treated here vs callbacks attached to the trainer? Asking since it came up in #5542

awaelchli commented:
@ananthsub The callbacks attached through this hook here will still go into the trainer as if they were added directly to the Trainer. 
Therefore, the same rules and limitations apply as for the Trainer. 
As a general rule of thumb, the callbacks that do have state should not change their class name (as this is the key by which they are saved to the checkpoint and retrieved) and there can only be one callback of each type. 
Callbacks that don't save state do not have these limitations, since they don't get saved.

awaelchli commented:
@PyTorchLightning/core-contributors would you find such a hook to configure model-specific callbacks directly in the LightningModule useful and if so which API would you prefer (3 choices above)?

rohitgr7 commented:
option 1 looks better to me since trainer callbacks can be configured/modified using `self.trainer.callbacks` if required.

> This was requested internally, see grid chat.

what's grid chat??

carmocca commented:
I agree with @akihironitta

s-rog commented:
While I would personally prefer 2 for maximum flexibility, I agree with Akihiro that it could be unintuitive for users.

teddykoker commented:
I think we can start with *1*, and we can always add *2* later if it becomes heavily requested

Borda commented:
@akihironitta @teddykoker @ananthsub @rohitgr7 @s-rog
 I have added a small poll ❤️ 🎉 🚀 on the description, so please check it there 🐰

Borda commented:
could it be a mix of the first two such that there will be an extra argument that would specify if the model CB is added as first or as last? ideally you would like to define some kind of the insert but it is maybe overkilled atm 🔢

SeanNaren commented:
I'd agree with option 1, with 2 being the future requirement for callbacks that interfere with each other!

awaelchli commented:
Hey all, thanks for voting :) 
I now implemented option 1 based on the feedback here and in slack. 
Additionally, I implemented the requirement proposed by @williamFalcon , that is, callbacks returned by the hook will replace callbacks with the same type. In this sense, the model has priority when it come to defining callbacks. 

See the test `test_attach_model_callbacks` for examles and how the two lists get merged.

rohitgr7 commented on pytorch_lightning/trainer/connectors/callback_connector.py:
```diff
@@ -107,3 +108,21 @@ def attach_model_logging_functions(self, model):
+        model_callbacks = model.configure_callbacks()
+        model_callback_types = [type(c) for c in model_callbacks]
+        # remove all callbacks with a type that occurs in model callbacks
+        all_callbacks = [c for c in self.trainer.callbacks if type(c) not in model_callback_types]
```
will it make sense to have some warning here?

awaelchli replied:
I would say definitely not a warning, because the way it is designed right now it is clearly a feature and the user first has to override the hook (and therefore read the docs) in order for all of this to take effect. 

One could argue maybe for a INFO or DEBUG level log statement

rohitgr7 replied:
INFO sound good

awaelchli replied:
I added that info. Let me know if you see any further improvements

rohitgr7 replied:
looks good.

rohitgr7 commented on pytorch_lightning/core/lightning.py:
```diff
@@ -990,6 +990,30 @@ def predict(self, batch: Any, batch_idx: int, dataloader_idx: Optional[int] = No
+        """
+        Configure model-specific callbacks.
+        When the model gets attached, e.g., when ``.fit()`` or ``.test()`` gets called,
+        the list returned here will be merged with the list of callbacks passed to the Trainer's `callbacks` argument.
```
```suggestion
        the list returned here will be merged with the list of callbacks passed to the Trainer's ``callbacks`` argument.
```

rohitgr7 commented on pytorch_lightning/core/lightning.py:
```diff
@@ -990,6 +990,30 @@ def predict(self, batch: Any, batch_idx: int, dataloader_idx: Optional[int] = No
+    def configure_callbacks(self):
+        """
+        Configure model-specific callbacks.
+        When the model gets attached, e.g., when ``.fit()`` or ``.test()`` gets called,
```
still don't know the use-case for `configure_callbacks` yet, but will it be good to have an argument similar to `stage` in the `setup` with `'fit'` and `'test'`?

carmocca replied:
I like that idea!

The use-case is supporting out-of-the-box LightningModules which require specific callbacks to work/train properly

awaelchli replied:
it makes sense. Would any of you like to add it as a follow up later? 

I'm hesitating to do it in this PR, because currently we have a bit of a confusing definition of "stages". There is now a "running_stage" and then there is the regular "stage" fit and test, and recently a _predict stage was introduced. 

Any thoughts?

rohitgr7 replied:
yes, in a separate one of course. We can just do it as it will be done with the setup here: https://github.com/PyTorchLightning/pytorch-lightning/issues/5658

Also, there is another way around it. We can just let users use `self.running_stage` (make it a property and return `trainer.running_stage`) and won't have any stage argument anywhere.

Although I'd still prefer having an argument, but just a thought.

awaelchli replied:
Hmm, if we don't add the stage arguments in this PR and decide to add it later, it will unfortunately be a breaking api change for the user.

Btw I tried the property approach here: #5724
However, it seems there is a bit of a mess going on hehe, it might be not so straightforward :)

ananthsub replied:
@rohitgr7 n00b question: why does the stage factor in here? The callback interface doesn't have the same pattern; it's just a flat list of events that the user can trigger on

awaelchli replied:
Ananth has a point. A callback that is only supposed to run say during test would just implement the test callback methods anyway.

rohitgr7 replied:
we do have some hooks that runs in both (fit and test) like `on_init_start/end`, `on_fit_start/end`.

ananthsub replied:
@rohitgr7 in that case should we add `stage` to those callback hooks directly?

carmocca commented:
> Also, I'm curious about the motivation for adding this feature...

Original reference: https://github.com/PyTorchLightning/pytorch-lightning/issues/3286#issuecomment-691254810

Borda commented:
@awaelchli how is it going here, still see you are skipping some tests :]

awaelchli commented on pytorch_lightning/trainer/connectors/callback_connector.py:
```diff
@@ -107,3 +108,21 @@ def attach_model_logging_functions(self, model):
+        """
+        model_callbacks = model.configure_callbacks()
+        model_callback_types = [type(c) for c in model_callbacks]
+        # remove all callbacks with a type that occurs in model callbacks
```
That could be an option. 
However, it would require additional logic when we allow to append without replacement, because we probably want to avoid keep adding callbacks when calling trainer.fit/test multiple times (not idempotent).

carmocca replied:
Good point. Lets hold off on it and see if people request it.

rohitgr7 commented on pytorch_lightning/trainer/connectors/callback_connector.py:
```diff
@@ -107,3 +108,29 @@ def attach_model_logging_functions(self, model):
+        override_types = model_callback_types.intersection(trainer_callback_types)
+        if override_types:
+            rank_zero_info(
+                "The following callbacks returned in LightningModule.configure_callbacks will override"
+                " existing callbacks attached to Trainer:"
```
```suggestion
                "The following callbacks returned in `LightningModule.configure_callbacks` will override"
                " existing callbacks passed to Trainer:"
```

rohitgr7 commented on tests/trainer/connectors/test_callback_connector.py:
```diff
@@ -0,0 +1,79 @@
+    with caplog.at_level(logging.INFO):
+        cb_connector.attach_model_callbacks(model)
+
+    assert "existing callbacks attached to Trainer: EarlyStopping, LearningRateMonitor" in caplog.text
```
```suggestion
    assert "existing callbacks passed to Trainer: EarlyStopping, LearningRateMonitor" in caplog.text
```

ananthsub commented on pytorch_lightning/trainer/connectors/callback_connector.py:
```diff
@@ -107,3 +108,29 @@ def attach_model_logging_functions(self, model):
             callback.log = model.log
             callback.log_dict = model.log_dict
+
+    def attach_model_callbacks(self, model: LightningModule) -> None:
```
is it meant to be a public interface for the connector?

awaelchli replied:
I thought I can leave it public because connectors are internal objects. But making it protected won't hurt :)

ananthsub commented on pytorch_lightning/core/lightning.py:
```diff
@@ -990,6 +990,30 @@ def predict(self, batch: Any, batch_idx: int, dataloader_idx: Optional[int] = No
         """
         return self(batch)
 
+    def configure_callbacks(self):
```
add a type-hint that this returns a list of callbacks?

awaelchli replied:
William once said he doesn't want type-hints in hooks api that  users override, the reason being that IDE's autocomplete it when overriding the hooks from the superclass. Then users would manually have to delete it. He sees this as not user friendly. So I'm not sure if I should add it :)

ananthsub commented on pytorch_lightning/trainer/connectors/callback_connector.py:
```diff
@@ -107,3 +108,29 @@ def attach_model_logging_functions(self, model):
+            A model which may or may not define new callbacks in
+            :meth:`~pytorch_lightning.core.lightning.LightningModule.configure_callbacks`.
+        """
+        model_callbacks = model.configure_callbacks()
+        model_callback_types = set(type(c) for c in model_callbacks)
```
you can short-circuit if model_callbacks empty

awaelchli replied:
luckily I had some wire here at home. Done!

ananthsub commented on pytorch_lightning/trainer/connectors/callback_connector.py:
```diff
@@ -107,3 +108,29 @@ def attach_model_logging_functions(self, model):
+                f" {', '.join(sorted(t.__name__ for t in override_types))}"
+            )
+        # remove all callbacks with a type that occurs in model callbacks
+        all_callbacks = [c for c in self.trainer.callbacks if type(c) not in override_types]
+        all_callbacks.extend(model_callbacks)
+        # TODO: connectors refactor: move callbacks list to connector and do not write Trainer state
+        self.trainer.callbacks = all_callbacks
```
i still think we need logic to ensure that any callbacks of type ModelCheckpoint run last. this problem will amplify if we have lightning module callbacks running after the trainer callbacks, though the risk exists today AFAICT. otherwise there's a risk that lightning module state is mutated in another callback after a checkpoint is saved, and that state is silently lost

awaelchli replied:
Yes I agree, and I remember it was mentioned several times in the past. Not sure why we haven't done it yet.
Adding a simple mapping here at the end should do it yes?

awaelchli replied:
See this [commit ](https://github.com/PyTorchLightning/pytorch-lightning/pull/5621/commits/607ac9fda21e5ed87474dbb41a277cee5b831d36)

Until now it seems we only made sure that the automatic/default checkpoint callback runs last. If the user defines their own callback list, it's not guaranteed. The changes I added will rectify that.

Perhaps I should brake this out into a separate PR

ananthsub replied:
> Perhaps I should brake this out into a separate PR

agreed!

tchaton replied:
Won't it be simpler to extend `model_callbacks` with `trainer_callbacks` ?

awaelchli replied:
so in the end all callbacks will be attached to the trainer. How the two are merged is simply a matter of definition. We can decide that. 
It could be 
`trainer.callbacks = trainer.callbacks + model_callbacks` or 
`trainer.callbacks = model_callbacks + trainer.callbacks`
both are possible. I don't think one is simpler than the other, we still need to let the callbacks in the model override the ones in trainer (according to william).

tchaton commented on pytorch_lightning/trainer/connectors/callback_connector.py:
```diff
@@ -107,3 +108,31 @@ def attach_model_logging_functions(self, model):
+            A model which may or may not define new callbacks in
+            :meth:`~pytorch_lightning.core.lightning.LightningModule.configure_callbacks`.
+        """
+        model_callbacks = model.configure_callbacks()
```
Can this merging function be a static method ?

awaelchli replied:
hmm, not so easy. need access to `self` to call `self._reorder_callbacks`

tchaton replied:
You could do `ModelConnector._reorder_callbacks`, having everything static :)

awaelchli replied:
done that

awaelchli commented on tests/trainer/connectors/test_callback_connector.py:
```diff
@@ -53,3 +73,69 @@ def test_all_callback_states_saved_before_checkpoint_callback(tmpdir):
+
+    # multiple callbacks of the same type in trainer
+    assert_composition(
+        trainer_callbacks=[LearningRateMonitor(),
+                           EarlyStopping(),
+                           LearningRateMonitor(),
+                           EarlyStopping()],
+        model_callbacks=[early_stopping, lr_monitor],
```
@akihironitta @Borda why does yapf format it like this? 
is it a bug?

Borda replied:
that looks strange... :/

awaelchli commented on tests/trainer/connectors/test_callback_connector.py:
```diff
@@ -53,3 +73,69 @@ def test_all_callback_states_saved_before_checkpoint_callback(tmpdir):
+
+    # multiple callbacks of the same type, in both trainer and model
+    assert_composition(
+        trainer_callbacks=[
+            LearningRateMonitor(), progress_bar,
+            EarlyStopping(),
+            LearningRateMonitor(),
+            EarlyStopping()
+        ],
```
same here, for some reason there are two elements on one line and the rest on the others. I thought it is supposed to give consistent formatting

Borda replied:
yes, this looks strange...

tchaton reviewed (approved):
LGTM !
